docs/proposals/loadbalancing.md

```diff
@@ -0,0 +1,338 @@
+
+```
+                Cloud                                  HaProxy pod
+                  ^                                         ^
+                  |                                         |
+            --------------------                     -------------------------
+           | GCE Controller pod |                   | Ha proxy Controller pod |
+            --------------------                     -------------------------
+                  |                                         |
+                  v                                         v
+            {Name: claim1,                          {Name: claim2,
+             Class: cloud,                           Class: bare-metal,
+             Ip: 172.11.14.111}                      Ip: 172.11.14.143:8080}
+                 /       \                              /        \
+             Path1      Path2                       Path3        Path4
+             type: l7   type: l7                    type: l4     type: l4
+             foo.com    bar.com                     foobar.com   baz.com
+             /a - svc1  /prod - svc2                cert: c1     cert: c2
+             /b - svc1                              svc4         svc5
+```
+
+#### Ingress Path
+
+This is an example ingress path. It encapsulates:
+* L7 proxying information (i.e map these urls to these backends)
+* Security configuration (i.e tlsMode, secret)
+* Metadata needed for more advanced loadbalancing (e.g session persistence)
+
+See [draft pr](https://github.com/kubernetes/kubernetes/pull/12825) for an actual description of the resource. See [alternatives](#alternatives) for why this deviates from existing implementations in OpenShift or GCE:
+
+```yaml
+apiVersion: v1
+kind: IngressPoint
+metadata:
+  name: l7ingress
+  type: l7
+spec:
+  host: example
+  tlsMode: Termination
+  secret: foosecret
+  pathMap:
+    "foo.example.com":
+    - url: "/foo/*"
+      service:
+        name: foosvc
+        namespace: default
+        port: 80
+    - url: "/bar/*"
+      service:
+        name: barsvc
+        namespace: default
+        port: 80
+    "bar.example.com"
+    - url: "/foobar/*"
+        service:
+          name: foobarsvc
+          namespace: default
+          port: 80
+status:
+  # include host here when we have DNS figured out.
+  address: 104.101.11.39
+```
+
+Put more rigidly:
+
+```go
+// IngressPoint represents a point for ingress traffic.
+type IngressPoint struct {
+	TypeMeta
+	ObjectMeta
+	Spec IngressPointSpec
+	Status IngressPointStatus
+}
+
+// IngressPointSpec describes the ingressPoint the user wishes to exist.
+type IngressPointSpec struct {
+	Host    string
+	IngressPoint map[string][]Path
+}
+
+// IngressPointStatus describes the current state of an ingressPoint.
+type IngressPointStatus struct {
+	Address string
+}
+
+// ServiceRef is a reference to a single service:port.
+type ServiceRef struct {
+	Name      string
+	Namespace string
+	Port      int64
```

Comment:
On Sep 14, 2015, at 3:07 PM, Prashanth B <notifications@github.com> wrote:

In docs/proposals/loadbalancing.md
<https://github.com/kubernetes/kubernetes/pull/12827#discussion_r39433549>:

> +// IngressPointSpec describes the ingressPoint the user wishes to exist.
> +type IngressPointSpec struct {
> +	Host    string
> +	IngressPoint map[string][]Path
> +}
> +
> +// IngressPointStatus describes the current state of an ingressPoint.
> +type IngressPointStatus struct {
> +	Address string
> +}
> +
> +// ServiceRef is a reference to a single service:port.
> +type ServiceRef struct {
> +	Name      string
> +	Namespace string
> +	Port      int64

Is there anything magic about the service that we really need to connect to?

Unfortunately now there is, the nodeport. Won't work on clouds without that
(or creating a seperate static ip per service, which is costly because
we're charging per ip).

Customers / user generally expected endpoints to be the target of the load
balancer.

I feel like we're already teaching people to think in terms of sets with
labels, and we already have a concept of sets of endpoints, the service,
why revert that?

There
are some weird cases - for instance, if with load balancing do I have to
add the port to my service in order to talk to an endpoint? Consider a pod
with two ports - one TLS and one HTTP. The TLS port is for internal
consumers, the HTTP ports are consumed by the load balancer because TLS is
imposed at the load balancer. Do I need to have HTTP port on the service?
Seems nice, but not truly required.

What is the common case here? I feel like we should optimize for that to
just work, and document the side cases. I had thought people would first
create a service, play around with it, then expose it to the world through
an ingresspoint. Something goes wrong you can always delete the
ingresspoint, and you still have an internal service to debug.



I'm talking more from the ingress ports perspective.  Services are nice,
but not required.  And most serious load balancers will go direct to
endpoints even if you are using node ports.  In fact, the load balancer
itself is the one making the decision, because it may be cluster internal
instead of external.

I'm *ok* with service as the indirection, I just don't think it's the happy
path once an LB is in real use.  Should LBs support dynamic reallocation of
ports when the service changes?  Is it up to the LB?  I think ingress can
assume endpoints always, and service for now, but I could see having
endpoints decoupled from services at some point.


—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/12827/files#r39433549>.